wandb/apis/public.py

```diff
@@ -5493,8 +5501,39 @@ def name(self):
     def _set_configure_launch_project(self, func):
         self.configure_launch_project = func
 
+    def _get_code_artifact(self, artifact_string):
+        artifact_string, _, is_id = util.parse_artifact_string(artifact_string)
+        if is_id:
+            code_artifact = Artifact.from_id(artifact_string, self._api._client)
+        else:
+            code_artifact = self._api.artifact(name=artifact_string, type="code")
+        if code_artifact is None:
+            raise LaunchError("No code artifact found")
+        return code_artifact
+
+    def _configure_launch_project_notebook(
+        self, launch_project, sourced_from_artifact=False
+    ):
+        notebook_info = self._job_info.get("source", {}).get("notebook")
+        assert notebook_info is not None
+        if not sourced_from_artifact:
+            artifact_string = notebook_info.get("notebook_artifact")
+            if artifact_string is None:
+                raise LaunchError(
+                    f"Job {self.name} notebook info missing source artifact"
+                )
+            code_artifact = self._get_code_artifact(artifact_string)
+            code_artifact.download(launch_project.project_dir)
+
+        new_fname = convert_jupyter_notebook_to_script(
+            "_session_history.ipynb", launch_project.project_dir
```

Comment:
If I understand right, all notebook runs have a code artifact with a _session_history, so even for git, we just use that. Its a bit tricky, but without the session history we dont actually run the same stuff as the original run (if the cells were run out of order). This is really a best effort approach, but I'd be fine with (and it would be simpler I think) to just run the notebook. It just adds another potential way for the run to differ from the original run

Reply:
Okay, I think that is true as long as the user has code saving enabled. So I guess as long as we make that clear in the docs it should be okay.